src/main/java/jenkins/branch/WorkspaceLocatorImpl.java

```diff
@@ -395,9 +397,9 @@ public void onDeleted(Item item) {
             }
             TopLevelItem tli = (TopLevelItem) item;
             Jenkins jenkins = Jenkins.get();
-            Computer.threadPoolForRemoting.submit(new CleanupTask(tli, jenkins));
             // Starts provisioner Thread which is tasked with starting cleanup Threads
-            new CleanupTaskProvisioner(tli, jenkins.getNodes()).run();
+            new CleanupTaskProvisioner(tli,
+                Stream.concat(Stream.of(jenkins), jenkins.getNodes().stream()).collect(Collectors.toList())).run();
```

Comment:
The comment here seems misleading given the code. I don't think there are any other threads involved. It looks like we synchronously call `run`, which synchronously calls `deleter.acquireThread()`, which synchronously calls `cleanupPool.acquire()`, so IIUC this code can cause `ItemListener.onDeleted` itself to block, which seems undesirable.

Maybe this went unnoticed previously because of the over-allocation you are trying to fix? Or perhaps I misunderstand. If my analysis is correct, then I worry that this PR might cause unexpected blockages, and bigger changes would be needed to fix things. For example, maybe we should instead all of these `CleanupTaskProvisioner` objects to a queue, and then have a single thread processing that queue and submitting the individual tasks to `Computer.threadPoolForRemoting` but making sure not to exceed whatever the desired limits are.

Reply:
Before the change, the listener would still possibly hang unless there is no Jenkins nodes. Now it might even if there is no Jenkins nodes. So you are right, in a way it is more likely to hang in that sense.

Another way to quickly fix the problem is to NOT do the `cleanupPool.release()` for the Controller built-in executor. Directly addresses the issue. But maybe not the more general inconsistent limit and the `ItemListener` contention.

> For example, maybe we should instead submit all of these CleanupTaskProvisioner objects to a queue, and then have a single thread processing that queue and submitting the individual tasks to Computer.threadPoolForRemoting but making sure not to exceed whatever the desired limits are.

Sounds good. Or was also wondering if `Computer.threadPoolRemoting` was required here ? Why not directly use a specific thread pool with unbounded queue (initial thread to 1, max number of threads to the limit set, 60s keep alive time) ?

Reply:
I have update the PR with a single thread executor service to manage the scheduling.